### PR TITLE
Implement LEDC based on ESP-IDF API

### LINKED ISSUE #5089: PWM ledCWrite(255) or ledcWrite(1023) is not 100% on
### Hardware:
Board: ESP32

### Description:

Unable to light off PWM common anode RGB Led when setting value to latest value of range.

Old known issue of ESP8266 [#5957](https://github.com/esp8266/Arduino/issues/5957) and now seen is [Tasmota](https://github.com/arendst/Tasmota/pull/11786) and my sample test program

The fix is described in comments of official sample [code](https://github.com/espressif/arduino-esp32/blob/404a31f4455d2459dad0b6533b7bfa70347e0044/libraries/ESP32/examples/AnalogOut/ledcWrite_RGB/ledcWrite_RGB.ino#L22)

But from logical working, having value going from 0 to 255 should means that 255 is full scale so always 1 and not having a little pulse. This does now works on ESP8266 with analogWrite(255) and comments have been updated to code according in this [PR](https://github.com/esp8266/Arduino/pull/7456/files)
```
// Per the Arduino docs at https://www.arduino.cc/reference/en/language/functions/analog-io/analogwrite/
// val: the duty cycle: between 0 (always off) and 255 (always on).
// So if val = 0 we have digitalWrite(LOW), if we have val==range we have digitalWrite(HIGH)
```
I know it's not arduino but I think if PWM could work as is `0 (always off) and max range value (always on)` would be less confusing

my `ledOff()`sample function with 256 does works, but should works with 255, much easier if we use an `uint8_t` as scale range.

```
void ledOff() {
  ledcWrite(LED_RED_CHAN, 256);
  ledcWrite(LED_GRN_CHAN, 256);
  ledcWrite(LED_BLU_CHAN, 256);
}
```

### PR DESCRIPTION
**Pull request by P-R-O-C-H-Y**
## Summary
This PR is refactoring of LEDC HAL in order to use IDF instead of current Register manipulation approach.
Fixing duty -> if all bits in resolution are set -> FULL ON

## Impact
None

## Related links
Fixed #5089

### CODE REVIEW

File: cores/esp32/esp32-hal-ledc.c [hunk 1]
```diff
@@ -72,223 +52,89 @@ xSemaphoreHandle _ledc_sys_lock = NULL;
...
-    LEDC_MUTEX_UNLOCK();
-}
+    ledc_timer_config_t ledc_timer = {
+        .speed_mode       = group,
+        .timer_num        = timer,
+        .duty_resolution  = bit_num,
+        .freq_hz          = freq,
+        .clk_cfg          = LEDC_DEFAULT_CLK
+    };
+    ESP_ERROR_CHECK(ledc_timer_config(&ledc_timer));
```
Comment (me-no-dev) [inline cores/esp32/esp32-hal-ledc.c:73]: do not use ESP_ERR_CHECK in Arduino. It uses ESP_LOGx inside which we can not control in Arduino

Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-ledc.c]: done

**Comment by HubertHuber** [comment]
ERROR when setting FREQUENCY to less than 1 Hz ---> IntegerDivideByZero

**Comment by me-no-dev** [comment]
ref: https://github.com/espressif/arduino-esp32/issues/6550

**Pull request merged**